src/trace/ord/merge_batcher/mod.rs

```diff
@@ -196,42 +220,68 @@ impl<D: Ord, R: MonoidValue> MergeSorter<D, R> {
         while self.queue.len() >= 2 {
             let list1 = self.queue.pop().unwrap();
             let list2 = self.queue.pop().unwrap();
+
             let merged = self.merge_by(list1, list2);
-            self.queue.push(merged);
+            if !merged.is_empty() {
+                self.queue.push(merged);
+            }
         }
 
         if let Some(mut last) = self.queue.pop() {
-            swap(&mut last, target);
+            // TODO: Reuse the `last` buffer somehow
+            if target.capacity() == 0 {
+                target.append(&mut last);
+            } else {
+                *target = last;
+            }
         }
         debug_assert!(self.queue.is_empty());
     }
 
     // merges two sorted input lists into one sorted output list.
-    #[inline(never)]
-    fn merge_by(&mut self, list1: Vec<Vec<(D, R)>>, list2: Vec<Vec<(D, R)>>) -> Vec<Vec<(D, R)>> {
+    // TODO: Split this into two functions:
+    //       - `fn merge_batch(Vec<(D, R)>, Vec<(D, R)>) -> Vec<(D, R)>`
+    //       - `fn merge_batches(Vec<Vec<(D, R)>>, Vec<Vec<(D, R)>>) -> Vec<Vec<(D,
+    //         R)>>`
+    // TODO: There's some things that need to be evaluated/tested/benchmarked here:
+    //       - When getting a new head1/head2 after the inner merge happens, is it
+    //         beneficial to pop from the opposite list (list1/list2) when one of
+    //         them is exhausted?
+    //       - Sometimes we can keep on merging into the same results vec when one
```

Comment:
Is this bullet point still valid?  I think we keep pushing to result even as we move `head` to the next element in the corresponding list.